File: HW/Programs/hw3.py
# Programming Language used : Python

def input_numbers():
    a, b = map(int, input("Enter two non-zero positive integers: ").split())
    
    if a <= 0 or b <= 0:
        print("Your inputs are not valid! Please enter two non-zero positive integers.")
    else:
        input_range = list(range(a, b+1))
        odd = []
        even = []
        
        #indirect function call using lambda functions
        seperate_function(lambda x: even.append(x) if x % 2 == 0 else odd.append(x), input_range)
        print("odd numbers: ", odd)
        print("Even numbers: ", even)
        
        operation = input("Enter operation (add_even, add_odd or, add_all): ")
        if(operation == "add_even" or operation == "add_odd" or operation == "add_all"):
            result = return_calculation(operation, even, odd)
            print("Result of operation " + str(operation) + " is: " + str(result))
        else:
            print("Invalid operation!")
    
# seperate_function calls another function (my_function) indirectly
def seperate_function(my_function, numbers):
    for element in numbers:
        my_function(element)
        
def return_calculation(operation, odd, even):
    
    #function calls within another function(indirect)
    if operation == "add_even":
        return add(even)
    elif operation == "add_odd":
        return add(odd)
    elif operation == "add_all":
        return add(even) + add(odd)
    else:
        return "Invalid operation!"

def add(numbers):
    total = 0
    for element in numbers:
        total += element
    return total

input_numbers()

File: HW/Programs/hw3.c
// Program is in C
#include <stdio.h>

// Function for even numbers (called indirectly  every time)
void even_func(int start, int end) {
    for (int i = start; i <= end; i++) {
        if (i % 2 == 0) {
            printf("%d ", i);
        }
    }
}

// Function for odd numbers (called indirectly every time)
void odd_func(int start, int end) {
    for (int i = start; i <= end; i++) {
        if (i % 2 != 0) {
            printf("%d ", i);
        }
    }
}

// Function to add even numbers (called indirectly accourind to the condition)
void add_even(int start, int end) {
    int sum = 0;
    for (int i = start; i <= end; i++) {
        if (i % 2 == 0) {
            sum += i;
        }
    }
    printf("\nSum of even numbers in the range: %d", sum);
}

// Function to add odd numbers (called indirectly accourind to the condition)
void add_odd(int start, int end) {
    int sum = 0;
    for (int i = start; i <= end; i++) {
        if (i % 2 != 0) {
            sum += i;
        }
    }
    printf("\nSum of odd numbers in the range: %d", sum);
}

int main() {
    int start, end;
    int (*indirect_func)(int, int); // Function pointer (to point to other functions for indirect calls)

    // Get input range
    printf("Please enter the starting and ending values of the range in seperate inputs\n");
    printf("\nEnter the starting number in the range: ");
    scanf("%d", &start);
    printf("\nEnter the end number in the range: ");
    scanf("%d", &end);

    // Choose processing function based on user input
    printf("\nEnter type of numbers to process (1: add all even numbers, 2: add all odd numbers): ");
    int choice;
    scanf("%d", &choice);
    
    //indirectly calling even handling function using indirect_func
    indirect_func = even_func;
    printf("\nEven numbers in the range:");
    indirect_func(start, end);
    
    //indirectly calling odd handling function using indirect_func
    indirect_func = odd_func;
    printf("\nOdd numbers in the range:");
    indirect_func(start, end);

    // if choice is even, indirectly call the function to handle add even 
    if (choice == 1) {
        indirect_func = add_even;
    } else if (choice == 2) {
        indirect_func = add_odd; // if choice is odd, indirectly call the function to handle odd odd
    } else {
        printf("\nInvalid choice! (enter 1: for even addition. 2: for odd additon.)");
        return 1; 
    }

    // Indirect calls using pointers (according to the input condition)
    indirect_func(start, end);

    return 0;
}
